### PR TITLE
First step of a JsonReader that reads a DOM.

### PR DESCRIPTION
**Pull request by swankjesse**
https://github.com/square/moshi/issues/89

### CODE REVIEW

File: moshi/src/main/java/com/squareup/moshi/ObjectJsonReader.java [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+  private boolean lenient;
+  private boolean failOnUnknown;
+
+  public ObjectJsonReader(Object root) {
+    scopes[stackSize] = JsonScope.NONEMPTY_DOCUMENT;
+    stack[stackSize++] = root;
+  }
+
+  @Override public void setLenient(boolean lenient) {
+    this.lenient = true;
```
Comment (serj-lotutovici) [inline moshi/src/main/java/com/squareup/moshi/ObjectJsonReader.java]: should be `this.lenient = lenient;`

Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/ObjectJsonReader.java]: Oh, ouch. Fixed!

File: moshi/src/main/java/com/squareup/moshi/ObjectJsonReader.java [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+
+/**
+ * This class reads a JSON document by traversing a Java object composing maps, lists, and JSON
```
Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/ObjectJsonReader.java]: nit: comprise, not compose 😀

File: moshi/src/test/java/com/squareup/moshi/JsonReaderPathTest.java [hunk 3]
```diff
@@ -172,7 +207,9 @@
   }
 
   @Test public void multipleTopLevelValuesInOneDocument() throws IOException {
-    JsonReader reader = newReader("[][]");
+    if (factory == Factory.JSON_OBJECT) return; // Object readers can't have multiple values.
```
Comment (JakeWharton) [inline moshi/src/test/java/com/squareup/moshi/JsonReaderPathTest.java]: could be `assumeTrue(factory != Factory.JSON_OBJECT);`

File: moshi/src/test/java/com/squareup/moshi/ObjectJsonReaderTest.java [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.fail;
+
+public final class ObjectJsonReaderTest {
+  @Test public void array() throws Exception {
+    List<Object> root = new ArrayList<>();
+    root.add("s");
+    root.add(1.5);
```
Comment (JakeWharton) [inline moshi/src/test/java/com/squareup/moshi/ObjectJsonReaderTest.java]: nit: missing `d`. other tests have it

File: moshi/src/test/java/com/squareup/moshi/JsonReaderPathTest.java (hunk 3 above)
Comment (swankjesse) [inline moshi/src/test/java/com/squareup/moshi/JsonReaderPathTest.java]: Dig.

File: moshi/src/test/java/com/squareup/moshi/ObjectJsonReaderTest.java (hunk 4 above)
Comment (swankjesse) [inline moshi/src/test/java/com/squareup/moshi/ObjectJsonReaderTest.java]: Fixed.

**Pull request merged**